doc/design/sql_program_dependency_analyze.md

```diff
@@ -0,0 +1,130 @@
+
+To achieve this, we need to analyze the hole SQL program and find out what tables are manipulated
+by each statement. One statement can read, write or do both on some tables, while other statements
+can also read, write the same table, in the above case, statement 1 is reading table `user_visits`
+and writing table `train_table`, statement 2,3,4,5 are reading table `train_table`, so they must
+start only when statement 1 is finished.
+
+We can do this by following below steps:
+
+1. Parse every SQL statement, return a map recording read/write type of the tables that the current
+   statement is manipulating. We'll get a list of this map in order, representing each SQL statement.
+   This will be done in the third-party parsers.
+2. Parse the SQLFlow extended SQL if there are any, return a similar map recording read/write type
+   of the models of the current statement (in our Go parser).
+3. Construct a graph using the following rules:
+    1. The statements are nodes in the graph, they should run in the original order.
+    2. If one statement reads a table/model, then construct one graph edge from statements that write
+       the table to the current statement.
+4. Execute this graph by submitting the graph as an Argo/Tekton workflow, each step is one single
+   statement.
+
+
+## Hazard
+
+SQL program is just like normal computer programs. We treat tables as variables in our case.
+Analyzing computer programs always have hazards: https://en.wikipedia.org/wiki/Hazard_(computer_architecture). In the example above, we can construct a graph like below:
+
+<p align="center">
+<img src="figures/hazard.png">
+</p>
+
+
+Note that the last "DROP TABLE" statement must execute after the two "Explain" statements, because
+the table is used by the explain statements before it could be changed (in this case, deleted). This
+case is called "Write After Read". To solve this kind of data hazards, we can add one dependency node
+"table WAR" between the explain statements and drop statement:
+
+<p align="center">
+<img src="figures/hazard_solve.png">
+</p>
+
+We also need to solve "Write After Write" hazard just like "Write After Read" if the SQL program has
+this situation.
+
+## Implementation
+
+In order to construct the dependency graph of the SQL program, we need to analyze the parsed SQL
+statements in every parser, including MySQL parser, Hive parser, calcite parser and SQLFlow extended
+parser. Since the Hive parser and calcite parser is written in Java, we need to pass the table read/write
+information from Java to Go:
+
+```proto
+message TableRW {
+  string table_name = 1;
+  // read: rw_type = 0 
+  // write: rw_type = 1
+  // both : rw_type = 2
+  int32 rw_type = 2;
+}
+
+message TableRWList {
+  repeated TableRW table_rw_list = 1;
+}
+
+message ParserResponse {
+  repeated string sql_statements = 1;
+  int32 index = 2;
+  string error = 3;
+  // Add table_rw field
+  repeated TableRWList table_rw = 4;
```

Comment:
```protobuf
repeated string input_tables = 4;
repeated string output_tables = 5;
repeated string input_models = 6;
repeated string output_models = 7;
```

Reply:
Consider the following case where `db.table` and `table` refers to the same table. The parser doesn't know this equivalence.  How could we infer it?

```sql
USE db;
CREATE TABLE table AS ...
SELECT * FROM table TO TRAIN ...
```

Reply:
We do not need `input_models` and `output_models` in the parser gRPC server, model read/write analyzing are in Go since the extended parser is written in Go.

Reply:
> Consider the following case where db.table and table refers to the same table. The parser doesn't know this equivalence. How could we infer it?

Added a section to describe the solution.